bpo-29847: Path subclasses raise TypeError given kwargs

jsaporta opened the pull request:
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->
TypeError raised in the subclasses as per the discussion on bugs.python.org.

<!-- issue-number: [bpo-29847](https://bugs.python.org/issue29847) -->
https://bugs.python.org/issue29847
<!-- /issue-number -->

jsaporta committed:
Issue 29847 fix with tests.

eamanu commented on Lib/pathlib.py:
```diff
@@ -1512,13 +1512,23 @@ class PosixPath(Path, PurePosixPath):
 
+    def __new__(cls, *args, **kwargs):
+        if kwargs:
+            raise TypeError("keyword arguments provided but ignored")
```
What about start the raise with uppercase?

jsaporta replied:
Other errors I've seen start with lowercase letters.
```python
>>> 3 + 'string'
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unsupported operand type(s) for +: 'int' and 'str'
```

eamanu commented on Lib/pathlib.py:
```diff
@@ -1512,13 +1512,23 @@ class PosixPath(Path, PurePosixPath):
     __slots__ = ()
 
+    def __new__(cls, *args, **kwargs):
+        if kwargs:
```
This weird for me because if you don't have kwargs on ```__new__(cls, *args, **kwargs)```
because this will raise a exception, what are you sending on ```super().__new__(cls, *args, **kwargs)```

jsaporta replied:
Not sure I understand you clearly, but if `kwargs` is an empty dictionary, then it will act as a falsy value in the if statement, and the exception will not be raised.

jdemeyer committed:
bpo-36974: inherit the vectorcall protocol (GH-13498)

jdemeyer committed:
bpo-36974: remove _PyObject_HasFastCall (GH-13460)

ZackerySpytz committed:
bpo-20602: Do not clear sys.flags and sys.float_info during shutdown (GH-8096)

There is no need to clear these immutable objects during shutdown.

tirkarthi committed:
bpo-37099: Silence DeprecationWarning in test_inspect (GH-13679)



Fix DeprecationWarning introduced in aee19f54f6fe45f6b3c906987941e5a8af4468e9


https://bugs.python.org/issue37099

jdemeyer committed:
bpo-36974: rename _FastCallKeywords -> _Vectorcall (GH-13653)

brettcannon committed:
Revert adding @maxking to CODEOWNERS file (GH-13660)



This reverts commit 71dc7c5fbd856df83202f39c1f41ccd07c6eceb7. Turns out you must have write access for CODEOWNERS to work.

agronholm committed:
bpo-36999: Add asyncio.Task.get_coro() (GH-13680)



https://bugs.python.org/issue36999

tirkarthi committed:
Fix audit event typo : urllib.request -> urllib.Request (GH-13550)



As per the PEP and the [audit event raised](https://github.com/python/cpython/blob/13d4e6a4a090031f8214e058ed3c8fd47767e05f/Lib/urllib/request.py#L524) in urllib.request this should be `urllib.Request`

cc: @zooba

andrewnc committed:
bpo-5028: fix doc bug for tokenize (GH-11683)



https://bugs.python.org/issue5028

ztane committed:
bpo-30969: Fix docs about the comparison in absence of __contains__ (GH-2761)

xdegaye committed:
bpo-36342: Fix test_multiprocessing in test_venv (GH-12513)



when platform lacks a functioning sem_open implementation



https://bugs.python.org/issue36342

Commit:
bpo-36953: Delay removal of ABCs from collections. (GH-13409)



Bump the removal to 3.9, indicate collections.abc available since 3.3,
replace version-changed directive to deprecated-removed.


https://bugs.python.org/issue36953

asottile committed:
bpo-5028: Fix up rest of documentation for tokenize documenting line (GH-13686)



https://bugs.python.org/issue5028

ilevkivskyi committed:
bpo-29262: Add get_origin() and get_args() introspection helpers to typing (GH-13685)

This is an old feature request that appears from time to time. After a year of experimenting with various introspection capabilities in `typing_inspect` on PyPI, I propose to add these two most commonly used functions: `get_origin()` and `get_args()`. These are essentially thin public wrappers around private APIs: `__origin__` and `__args__`.

As discussed in the issue and on the typing tracker, exposing some public helpers instead of `__origin__` and `__args__` directly will give us more flexibility if we will decide to update the internal representation, while still maintaining backwards compatibility.

The implementation is very simple an is essentially a copy from `typing_inspect` with one exception: `ClassVar` was special-cased in `typing_inspect`, but I think this special-casing doesn't really help and only makes things more complicated.

jdemeyer committed:
bpo-36974: tp_print -> tp_vectorcall_offset and tp_reserved -> tp_as_async (GH-13464)

Automatically replace
tp_print -> tp_vectorcall_offset
tp_compare -> tp_as_async
tp_reserved -> tp_as_async

serhiy-storchaka committed:
bpo-339827: Do not swallow exceptions in the _ssl module. (GH-12756)

serhiy-storchaka committed:
bpo-36548: Improve the repr of re flags. (GH-12715)

terryjreedy committed:
IDLE - Capitalize search dialogs' 'Close' button label. (#13691)

It seems to be the only widget label not capitalized.

serhiy-storchaka committed:
bpo-31829: Make protocol 0 pickles be loadable in text mode in Python 2. (GH-11859)

Escape ``\r``, ``\0`` and ``\x1a`` (end-of-file on Windows) in Unicode strings.

serhiy-storchaka committed:
bpo-26660, bpo-35144: Fix permission errors in TemporaryDirectory cleanup. (GH-10320)

TemporaryDirectory.cleanup() failed when non-writeable or non-searchable
files or directories were created inside a temporary directory.

tiran committed:
bpo-34271: Add ssl debugging helpers (GH-10031)

The ssl module now can dump key material to a keylog file and trace TLS
protocol messages with a tracing callback. The default and stdlib
contexts also support SSLKEYLOGFILE env var.

The msg_callback and related enums are private members. The feature
is designed for internal debugging and not for end users.

Signed-off-by: Christian Heimes <christian@python.org>

ZackerySpytz committed:
bpo-36379: __ipow__ must be a ternaryfunc, not a binaryfunc (GH-13546)



If a type's __ipow__ method was implemented in C, attempting to use
the *modulo* parameter would cause crashes.


https://bugs.python.org/issue36379

pablogsal committed:
bpo-37108: Support super with methods that use positional-only arguments (GH-13695)

pablogsal committed:
Update data model docs to include missing attributes for code objects (GH-13696)

Include and document co_posonlyargcount and co_kwonlyargcount

pablogsal committed:
bpo-37112: Allow compile to work on AST with positional only arguments with defaults (GH-13697)

pablogsal committed:
bpo-37115: Support annotations in positional-only arguments (GH-13698)

MakDon committed:
bpo-37094: Add example for TestCase.skipTest in unittest doc (GH-13645)



Also includes other minor test skipping doc improvements.


https://bugs.python.org/issue37094

tiran committed:
bpo-26835: Add file sealing constants to fcntl (GH-13694)

Co-authored-by: nanjekyejoannah <nanjekyejoannah@gmail.com>

akshaysharma096 committed:
bpo-25735: math.factorial doc should mention integer return type (GH-6420)

ZackerySpytz committed:
bpo-12639: msilib.Directory.start_component() fails if *keyfile* is not None (GH-13688)

msilib.Directory.start_component() was passing an extra argument to CAB.gen_id().

pablogsal committed:
Document changes for PyCode_New regarding PEP570 (GH-13706)

pablogsal committed:
bpo-26826: Expose copy_file_range in the os module (GH-7255)

ammaraskar committed:
bpo-33361: Fix bug with seeking in StreamRecoders (GH-8278)

csabella committed:
bpo-15115: Document deprecation of email.encoders in Python 3 (GH-5354)

eamanu committed:
bpo-37105: Add deprecated-remove information on stream doc (#13672)

* Add deprecated-remove information on stream doc

According to the code on streams.py the functions:
``open_connection()``, ``start_server()``, ``open_unix_connection()``,
``start_unix_server()`` are deprecated. I infor that on
documentation.

ZackerySpytz committed:
bpo-12202: Properly check MsiSummaryInfoGetProperty() calls in msilib (GH-13711)

tim-one committed:
bpo-37029:  keep usable_arenas in sorted order without searching (#13612)

This adds a vector of "search fingers" so that usable_arenas can be kept in sorted order (by number of free pools) via constant-time operations instead of linear search.

This should reduce worst-case time for reclaiming a great many objects from O(A**2) to O(A), where A is the number of arenas.  See bpo-37029.

ericsnowcurrently committed:
bpo-36818: Add PyInterpreterState.runtime field. (gh-13129)

https://bugs.python.org/issue36818

timhoffm committed:
Improve docstring of list.sort (GH-8516)

rkday committed:
bpo-29984: Improve 'heapq' test coverage (GH-992)

mariocj89 committed:
Add option to trace to run modules (GH-5134)

Adds a new option in trace that allows tracing runnable modules. It is
exposed as `--module module_name` as `-m` is already in use for another
argument.

Windsooon committed:
bpo-18911: clarify that the minidom XML writer receives texts but not bytes (GH-13352)

FR4NKESTI3N committed:
bpo-35431: Implemented math.comb (GH-11414)

serhiy-storchaka committed:
bpo-37116: Use PEP 570 syntax for positional-only parameters. (GH-13700)

serhiy-storchaka committed:
Use more PEP 570 syntax in the documentation. (GH-13720)

bharel committed:
bpo-36813: Fix QueueListener to call task_done() upon termination. (GH-13113)



Fixed QueueListener in order to avoid random deadlocks.
Unable to add regression tests atm due to time constraints, will add it in a bit.
Regarding implementation, although it's nested, it does not cause performance issues whatsoever, and does not call task_done() in case of an exception (which is the right thing to do IMHO).


https://bugs.python.org/issue36813

mdickinson committed:
Move whats-new entry for math.factorial to the math module section. (GH-13723)

matrixise committed:
Doc: Correct the creation year and the credits of the Logo Programming language (GH-13520)

tonybaloney committed:
bpo-31968: Documentation -- add clarification on the globals dict for exec() (GH-13140)

pablogsal committed:
bpo-37122: Make co->co_argcount represent the total number of positonal arguments in the code object (GH-13726)

franciscouzo committed:
Improve exception message for str.format (GH-12675)

serhiy-storchaka committed:
bpo-35431: Refactor math.comb() implementation. (GH-13725)

* Fixed some bugs.
* Added support for index-likes objects.
* Improved error messages.
* Cleaned up and optimized the code.
* Added more tests.

evanunderscore committed:
bpo-28595: Allow shlex whitespace_split with punctuation_chars (GH-2071)

scoder committed:
Clean up and reduce visual clutter in the makeunicode.py script. (GH-7558)

pablogsal committed:
Fix compiler warnings in the pystrehex module (GH-13730)

pablogsal committed:
bpo-36842: Pass positional only parameters to code_new audit hook (GH-13707)

sir-sigurd committed:
bpo-34303: Micro-optimizations in functools.reduce() (GH-8598)

csabella committed:
bpo-32411: IDLE: Remove line number sort in browser.py (#5011)

Insertion in line order makes sorting keys by line order unneeded.

serhiy-storchaka committed:
bpo-20092. Use __index__ in constructors of int, float and complex. (GH-13108)

ZackerySpytz committed:
Fix the error handling in bytesio_sizeof(). (GH-10459)

bytesio_sizeof() must check if an error has occurred in _PySys_GetSizeOf().

marco-buttu committed:
bpo-29414: Change 'the for statement is such an iterator' in Tutorial (GH-273)

ericsnowcurrently committed:
bpo-33608: Factor out a private, per-interpreter _Py_AddPendingCall(). (gh-13714)

rhettinger committed:
Put math.comb() docs is correct place alphabetically (GH-13734)

asottile committed:
Improve version added references in `typing` module docs (GH-13457)

serhiy-storchaka committed:
Add more tests for preserving identity in marshal. (GH-13736)

serhiy-storchaka committed:
bpo-37128: Add math.perm(). (GH-13731)

mdickinson committed:
bpo-36027: Extend three-argument pow to negative second argument (GH-13266)

asvetlov committed:
bpo-35621: Support running subprocesses in asyncio when loop is executed in non-main thread (#13630)

pablogsal committed:
bpo-37126: Allow structseq objects to be tracked by the GC (GH-13729)

timhoffm committed:
Add description to the command line help of the argument clinic (GH-8518)

csabella committed:
bpo-35610: IDLE - Replace .context_use_ps1 with .prompt_last_line (GH-11307)

Changes in bpo- 31858 made the less informative 'context_use_ps1' redundant.

pablogsal committed:
bpo-37124: Fix reference leak in test_msilib (GH-13750)

pablogsal committed:
Call PyObject_GC_UnTrack in structseq dealloc (GH-13751)

mangrisano committed:
bpo-37014: Update docstring and Documentation of fileinput.FileInput(). (GH-13545)

* bpo-37014: Update docstring and Documentation of fileinput.FileInput()

* Explain the behavior of fileinput.FileInput() when reading stdin.

* Update blurb.

* bpo-37014: Fix typo in the docstring and documentation.

vstinner committed:
bpo-36829: test_threading: Fix a ref cycle (GH-13752)

encukou committed:
test_gdb.test_pycfunction: test more calling conventions (GH-13668)

As the code paths for various METH_* conventions are diverging due
to optimizations, we should check they continue to be covered by
GDB integration.

mangrisano committed:
bpo-19184: Update the documentation of dis module. (GH-13652)

* bpo-19184: Update the documentation of dis module

* Explain the behavior of the number of arguments of RAISE_VARGARGS
  opcode.

* bpo-19184: Update blurb.

* bpo-19184: Fix typo in the dis Documentation.

* bpo-19184: Address review comments and improve the doc

* bpo-19184: Remove news file.

encukou committed:
bpo-36974: Make tp_call=PyVectorcall_Call work for inherited types (GH-13699)

When inheriting a heap subclass from a vectorcall class that sets
`.tp_call=PyVectorcall_Call` (as recommended in PEP 590), the subclass does
not inherit `_Py_TPFLAGS_HAVE_VECTORCALL`, and thus `PyVectorcall_Call` does
not work for it.

This attempts to solve the issue by:
* always inheriting `tp_vectorcall_offset` unless `tp_call` is overridden
  in the subclass
* inheriting _Py_TPFLAGS_HAVE_VECTORCALL for static types, unless `tp_call`
  is overridden
* making `PyVectorcall_Call` ignore `_Py_TPFLAGS_HAVE_VECTORCALL`

This means it'll be ever more important to only call `PyVectorcall_Call`
on classes that support vectorcall. In `PyVectorcall_Call`'s intended role
as `tp_call` filler, that's not a problem.

asvetlov committed:
bpo-35621: Fix tests when SafeChildWatcher is expected instead of ThreadedChildWatcher (GH-13754)



https://bugs.python.org/issue35621

encukou committed:
bpo-36027 bpo-36974: Fix "incompatible pointer type" compiler warnings (GH-13758)

tirkarthi committed:
Fix typos in docs and docstrings (GH-13745)

encukou committed:
bpo-37012: Clean up special cases in PyType_FromSpecWithBases slot assignments (GH-13496)

The main slot assignment loop is now if-else if ladder, making the
control flow clearer.

Based on suggestion by Victor Stinner in:
https://github.com/python/cpython/pull/10304/#issuecomment-491123026

vstinner committed:
bpo-37100: Fix test_coroutines with -Werror (GH-13756)

test_coroutines: test_unawaited_warning_when_module_broken() now uses
support.check_warnings() to catch expected RuntimeWarning.

ilevkivskyi committed:
Update the annotated assignment docs (GH-13757)

jdemeyer committed:
bpo-36974: document PEP 590 (GH-13450)

jdemeyer committed:
bpo-36974: add some assertions for PEP 590 (GH-13682)

encukou committed:
bpo-36027: Really fix "incompatible pointer type" compiler warning (GH-13761)

Apologies for the earlier hasty attempt.

Carreau committed:
bpo-36896: Clarify that some types constructors are unstable (GH-13271)

vstinner committed:
bpo-37069: tests use catch_unraisable_exception() (GH-13762)

Modify test_coroutines, test_cprofile, test_generators, test_raise,
test_ssl and test_yield_from to use
support.catch_unraisable_exception() rather than
support.captured_stderr().

test_thread: remove test_save_exception_state_on_error() which is now
updated. test_unraisable_exception() checks that sys.unraisablehook()
is called to handle _thread.start_new_thread() exception.

test_cprofile now rely on unittest for test discovery: replace
support.run_unittest() with unittest.main().

hackaugusto committed:
bpo-33569 Preserve type information with dataclasses.InitVar (GH-8927)

cclauss committed:
Fix variable name copy/paste error in build-installer.py (GH-13038)

rhettinger committed:
bpo-36546: Add design notes to aid future discussions (GH-13769)

tirkarthi committed:
IDLE: Fix typos in docs and comments (GH-13749)

pablogsal committed:
bpo-35814: Allow unpacking in r.h.s of annotated assignment expressions (GH-13760)

ned-deily committed:
bpo-36231:  Support building on macOS without /usr/include (GH-13773)

ned-deily committed:
Pin macOS installer Sphinx to v2.0.1 (GH-13774)

methane committed:
bpo-26219: per opcode cache for LOAD_GLOBAL (GH-12884)

This patch implements per opcode cache mechanism, and use it in
only LOAD_GLOBAL opcode.

Based on Yury's opcache3.patch in bpo-26219.

methane committed:
bpo-26219: remove unused code (GH-13775)

This code was for deoptimization, which is removed from
PR-12884.

vstinner committed:
Add credits to What's New in Python 3.8 (GH-13776)

* Credit myself and others.
* Complete asyncio changes.

JulienPalard committed:
Doc fix: duplicate object description of email.message (GH-13742)

Commit:
bpo-37087: Adding native ID support for OpenBSD (GH-13654)

vstinner committed:
bpo-37137: Fix test_asyncio: use TestCase.set_event_loop() (GH-13779)

Replace asyncio.set_event_loop() with TestCase.set_event_loop() of
test_asyncio.utils: this method calls TestCase.close_loop() which
waits until the executor completes, to avoid leaking dangling
threads.

Inherit from test_asyncio.utils.TestCase rather than
unittest.TestCase.

vstinner committed:
Revert "bpo-33608: Factor out a private, per-interpreter _Py_AddPendingCall(). (gh-13714)" (GH-13780)

This reverts commit 6a150bcaeb190d1731b38ab9c7a5d1a352847ddc.

tiran committed:
bpo-37081: Test with OpenSSL 1.1.1c (GH-13631)

Signed-off-by: Christian Heimes <christian@python.org>

tiran committed:
bpo-34271: Fix compatibility with 1.0.2 (GH-13728)

Fix various compatibility issues with LibreSSL and OpenSSL 1.0.2
introduced by bpo-34271.

Signed-off-by: Christian Heimes <christian@python.org>

tiran committed:
bpo-36868: Fix what's new for SSLContext.hostname_checks_common_name (GH-13248)



What's new now mentions SSLContext.hostname_checks_common_name instead of SSLContext.host_flags.


https://bugs.python.org/issue36868

tiran committed:
bpo-37120: Add SSLContext.num_tickets (GH-13719)

Signed-off-by: Christian Heimes <christian@python.org>

vstinner committed:
bpo-34037, asyncio: add BaseEventLoop.wait_executor_on_close (GH-13786)

Add BaseEventLoop.wait_executor_on_close attribute: true by default.

loop.close() now waits for the default executor to finish by default.
Set loop.wait_executor_on_close attribute to False to not wait for
the executor.

methane committed:
bpo-37146: disable opcache when Py_DEBUG is defined (GH-13787)

--with-pydebug is commonly used to find memory leaks.
But opcache makes it harder.
So disable opcache when Py_DEBUG is defined.

asvetlov committed:
Revert "bpo-35621: Support running subprocesses in asyncio when loop is executed in non-main thread (#13630)" (GH-13793)



https://bugs.python.org/issue35621

vstinner committed:
Revert "bpo-36818: Add PyInterpreterState.runtime field. (gh-13129)" (GH-13795)

This reverts commit 396e0a8d9dc65453cb9d53500d0a620602656cfe.

rhettinger committed:
bpo-35431: Drop the k <= n requirement (GH-13798)

pablogsal committed:
bpo-37148: Fix asyncio test that check for warning when running the test suite with huntleaks (GH-13800)

ilevkivskyi committed:
More updates to the annotated assignments docs (GH-13794)

rhettinger committed:
Fix grammar (GH-13801)

ambv committed:
Revert "bpo-34037, asyncio: add BaseEventLoop.wait_executor_on_close (GH-13786)" (#13802)

This reverts commit 0f0a30f4da4b529e0f7df857b9f575b231b32758.

asvetlov committed:
Make StreamServer.close() tests more robust (GH-13790)

asvetlov committed:
bpo-37142: Make asyncio stream tests more robust again (GH-13804)

Make test_stream_server_close() implementation following test_stream_server_abort().
 Add explicit timeout for tests that can hang.

vstinner committed:
bpo-26219: Fix compiler warning in _PyCode_InitOpcache() (GH-13809)

Fix this MSVC warning:

    objects\codeobject.c(264): warning C4244: '=':
    conversion from 'Py_ssize_t' to 'unsigned char', possible loss of data

vstinner committed:
bpo-36778: Remove outdated comment from CodePageTest (GH-13807)

CP65001Test has been removed.

mariocj89 committed:
bpo-30699: Improve example on datetime tzinfo instances (GH-4290)

* Improve example on tzinfo instances

Move from GMTX to TZX when naming the classes, as GMT1 might be rather
confusing as seen in the reported issue.
In addition, move to UTC over GMT and improve the tzname implementation.

* Simplify datetime with tzinfo example

Move the example in the documentation to just use timezone.utc and a
user defined Kabul timezone rather than having two user defined
timezones with DST.

Kabul timezone is still interesting as it changes its offset but not
based on DST. This is more accurate as the previous example was missing
information about the fold attribute. Additionally, implementing the fold
attribute was rather complex and probably not relevant enough for the
section "datetime with tzinfo".

Lothiraldan committed:
Fix extraneous whitespace in QueueListener.prepare (GH-13803)

zooba committed:
bpo-36742: Corrects fix to handle decomposition in usernames (#13812)

vstinner committed:
bpo-37153: test_venv.test_mutiprocessing() calls pool.terminate() (GH-13816)

test_venv.test_mutiprocessing() now explicitly calls pool.terminate()
to wait until the pool completes.

maxking committed:
bpo-35805: Add parser for Message-ID email header. (GH-13397)

* bpo-35805: Add parser for Message-ID header.

This parser is based on the definition of Identification Fields from RFC 5322
Sec 3.6.4.

This should also prevent folding of Message-ID header using RFC 2047 encoded
words and hence fix bpo-35805.

* Prevent folding of non-ascii message-id headers.
* Add fold method to MsgID token to prevent folding.

ambv committed:
Python 3.8.0b1

maxking committed:
bpo-30835: email: Fix AttributeError when parsing invalid CTE (GH-13598)

* bpo-30835: email: Fix AttributeError when parsing invalid Content-Transfer-Encoding

Parsing an email containing a multipart Content-Type, along with a
Content-Transfer-Encoding containing an invalid (non-ASCII-decodable) byte
will fail. email.feedparser.FeedParser._parsegen() gets the header and
attempts to convert it to lowercase before comparing it with the accepted
encodings, but as the header contains an invalid byte, it's returned as a
Header object rather than a str.

Cast the Content-Transfer-Encoding header to a str to avoid this.

Found using the AFL fuzzer.

Reported-by: Daniel Axtens <dja@axtens.net>
Signed-off-by: Andrew Donnellan <andrew@donnellan.id.au>

* Add email and NEWS entry for the bugfix.

PetterS committed:
bpo-35047: Update whatsnew/3.8 for better mock error message (GH-13746)

ambv committed:
Merge tag 'v3.8.0b1'

Python 3.8.0b1

ambv committed:
Bump to 3.9.0a0

zware committed:
Bump version in AppVeyor config (#13822)

zware committed:
bpo-34282: Remove deprecated enum _convert method (GH-13823)

JulienPalard committed:
Doc: Python 3.9 in sidebar and version switcher. (GH-13824)

terryjreedy committed:
bpo-35763: Make IDLE calltip note about '/' less obtrusive (GH-13791)

Add it to the end of the first line if there is room.  Tests were reworked.

eamanu committed:
bpo-36373:  Deprecate explicit loop parameter in all public asyncio APIs [streams] (GH-13671)



This PR deprecate explicit loop parameters in all public asyncio APIs

This issues is split to be easier to review.

Second step: streams.py





https://bugs.python.org/issue36373

bentyeh committed:
Fix documentation (GH-13721)



Based on the source code https://github.com/python/cpython/blob/4a686504eb2bbf69adf78077458508a7ba131667/Lib/multiprocessing/pool.py#L755 AsyncResult.successful() raises a ValueError, not an AssertionError.

ZackerySpytz committed:
bpo-34767: Do not always create a collections.deque() in asyncio.Lock() (GH-13834)



https://bugs.python.org/issue34767

vstinner committed:
bpo-26836: Document os.memfd_create() name parameter (GH-13838)



https://bugs.python.org/issue26836

rhettinger committed:
bpo-37158: Simplify and speed-up statistics.fmean() (GH-13832)

serhiy-storchaka committed:
[3.9] bpo-37116: Use PEP 570 syntax for positional-only parameters. (GH-12620)

Turn deprecation warnings added in 3.8 into TypeError.

maxking committed:
bpo-21315: Fix parsing of encoded words with missing leading ws. (#13425)

* bpo-21315: Fix parsing of encoded words with missing leading ws.

Because of missing leading whitespace, encoded word would get parsed as
unstructured token. This patch fixes that by looking for encoded words when
splitting tokens with whitespace.

Missing trailing whitespace around encoded word now register a defect
instead.

Original patch suggestion by David R. Murray on bpo-21315.

vstinner committed:
bpo-33725, multiprocessing doc: rephase warning against fork on macOS (GH-13841)

Co-Authored-By: Barry Warsaw <barry@python.org>

pablogsal committed:
bpo-37134: Add PEP570 notation to the documentation (GH-13743)

epicfaace committed:
bpo-35551: encodings update (GH-11446)

rhettinger committed:
bpo-37165: Convert _count_elements to the argument clinic (GH-13848)

pablogsal committed:
bpo-37134: Use PEP570 syntax for sum() (GH-13851)

warsaw committed:
Add importlib.metadata to what's new. (#13855)

methane committed:
bpo-35551: remove mac_centeuro encoding (GH-13856)

It is alias to mac_latin2 now.

doerwalter committed:
bpo-2661: Make mapping tests better usable for custom mapping classes. (GH-11157)



In test_fromkeys() the derived test class now supports all arguments in its
constructor so that the class to be tested can use its own constructor in its
fromkeys() implementation.

In test_mutatingiteration() the test fails as soon as iterating over a
dictionary with one entry and adding new entries in the loop iterates more
than once (to avoid endless loops in faulty implementations).



https://bugs.python.org/issue2661

vstinner committed:
bpo-36763, _testembed: enable assert() in release mode (GH-13857)

zooba committed:
bpo-37156: Fix libssl DLL tag in MSI sources (GH-13866)

websurfer5 committed:
bpo-36520: Email header folded incorrectly (#13608)

* bpo-36520: reset the encoded word offset when starting a new
line during an email header folding operation

* 📜🤖 Added by blurb_it.

* bpo-36520: add an additional test case, and provide descriptive
comments for the test_folding_with_utf8_encoding_* tests

* bpo-36520: fix whitespace issue

* bpo-36520: changes per reviewer request -- remove extraneous
backslashes; add whitespace between terminating quotes and
line-continuation backslashes; use "bpo-" instead of
"issue #" in comments

ZackerySpytz committed:
bpo-37170: Fix the cast on error in PyLong_AsUnsignedLongLongMask() (GH-13860)

pablogsal committed:
bpo-37134: Add PEP570 notation to the signature of byte{array}.translate (GH-13874)

taleinat committed:
bpo-37177: make IDLE's search dialogs transient (GH-13869)

This avoids the search dialogs being hidden behind the editor window.

eric-wieser committed:
bpo-37188: Fix a divide-by-zero in arrays of size-0 objects (#13881)

csabella committed:
IDLE: Standardize naming convention for DummyEditwin in tests (GH-13876)

* Change from Dummy_Editwin to DummyEditwin to match other tests.

skrah committed:
Use assertEqual(). (#13883)

jdemeyer committed:
bpo-37151: simplify classmethoddescr_call (GH-13340)

vstinner committed:
bpo-37169: Rewrite _PyObject_IsFreed() unit tests (GH-13888)

Replace two Python function calls with a single one to ensure that no
memory allocation is done between the invalid object is created and
when _PyObject_IsFreed() is called.

vstinner committed:
bpo-37191: Move TestPEP590 from test_capi to test_call (GH-13892)

paulmon committed:
bpo-37181: Fix test_regrtest failures on Windows arm64 (GH-13872)

jdemeyer committed:
bpo-37138: fix undefined behaviour with memcpy() on NULL array (GH-13867)

0x29a committed:
bpo-37150: Throw ValueError if FileType class object was passed in add_argument (GH-13805)



There is a possibility that someone (like me) accidentally will omit parentheses with `FileType` arguments after `FileType`, and parser will contain wrong file until someone will try to use it. 

Example:
```python
parser = argparse.ArgumentParser()
parser.add_argument('-x', type=argparse.FileType)
```


https://bugs.python.org/issue37150

asottile committed:
cross port importlib-metadata PR #76 (#13903)

https://gitlab.com/python-devs/importlib_metadata/merge_requests/76

flying-sheep committed:
bpo-37173: Show passed class in inspect.getfile error (GH-13861)



Currently, inspect.getfile(str) will report nonsense:

```pytb
>>> inspect.getfile(str)
TypeError: <module 'builtins' (built-in)> is a built-in class
```

This fixes that


https://bugs.python.org/issue37173

ammaraskar committed:
bpo-29505: Fix interpreter in fuzzing targets to be relocatable (GH-13907)

Commit:
bpo-34886: Fix subprocess.run handling of exclusive arguments (GH-11727)

Fix an unintended ValueError from :func:`subprocess.run` when checking for
conflicting `input` and `stdin` or `capture_output` and `stdout` or `stderr` args
when they were explicitly provided but with `None` values within a passed in
`**kwargs` dict rather than as passed directly by name.

rhettinger committed:
bpo-37178: Allow a one argument form of math.perm() (GH-13905)

n0npax committed:
bpo-11122: fix hardcoded path checking for rpmbuild in bdist_rpm.py (GH-10594)

ncoghlan committed:
Add some placeholder notes for major 3.8 features (GH-13927)

pitrou committed:
bpo-36785: PEP 574 What's New entry (#13931)

serhiy-storchaka committed:
Do not use explicit inheritance from object in the documentation. (GH-13936)

tiran committed:
bpo-37215: Fix dtrace issue introduce by bpo-36842 (GH-13940)



Signed-off-by: Christian Heimes <christian@python.org>


https://bugs.python.org/issue37215

vstinner committed:
bpo-18748: Fix _pyio.IOBase destructor (closed case) (GH-13952)

_pyio.IOBase destructor now does nothing if getting the closed
attribute fails to better mimick _io.IOBase finalizer.

vstinner committed:
bpo-37223: test_io: silence destructor errors (GH-13954)

Implement also MockNonBlockWriterIO.seek() method.

aaronpaulhurst committed:
closes bpo-35184: Fix XML_POOR_ENTROPY option that breaks makesetup parsing of pyexpat line in Setup. (GH-13064)

When the line is uncommented, the equals character causes it to be incorrectly interpreted
as a macro definition by makesetup.  This results in invalid Makefile output.

The expat code only requires XML_POOR_ENTROPY to be defined; the value is unnecessary.

rhettinger committed:
bpo-37219: Remove erroneous optimization for differencing an empty set (GH-13965)

asvetlov committed:
bpo-36607: Eliminate RuntimeError raised by asyncio.all_tasks() (GH-13971)



If internal tasks weak set is changed by another thread during iteration.



https://bugs.python.org/issue36607

benjaminp committed:
closes bpo-33758: Skip test_get_type_hints_modules_forwardref. (GH-13977)

This test "works" if things are run in the right order, so it's better to use @skip than @expectedFailure here.

benjaminp committed:
Fix test_posix if RWF_HIPRI is defined but not preadv2. (GH-13980)

If preadv2 is not available, preadv will raise NotImplementedError.

gvanrossum committed:
bpo-35766: What's new in the ast and typing modules (#13984)

paulmon committed:
bpo-37238: Enable building for Windows using Visual Studio 2019 (GH-13988)

gvanrossum committed:
bpo-35766: Change format for feature_version to (major, minor) (GH-13992)



(A single int is still allowed, but undocumented.)



https://bugs.python.org/issue35766

vstinner committed:
bpo-35766: compile(): rename feature_version parameter (GH-13994)

Rename compile() feature_version parameter to _feature_version and
convert it to a keyword-only parameter.

Update also test_type_comments to pass feature_version as a tuple.

vstinner committed:
bpo-36918: Fix "Exception ignored in" in test_urllib (GH-13996)

Mock the HTTPConnection.close() method in a few unit tests to avoid
logging "Exception ignored in: ..." messages.

vstinner committed:
bpo-26219: Fix compiler warning in _PyCode_InitOpcache() (GH-13997)

Fix MSVC warning:

    objects\codeobject.c(285): warning C4244: '=':
    conversion from 'Py_ssize_t' to 'unsigned char',
    possible loss of data

laike9m committed:
bpo-32625: Updated documentation for EXTENDED_ARG. (GH-13985)

Python 3.6 changed the size of bytecode instruction, while the documentation for `EXTENDED_ARG` was not updated accordingly.

ammaraskar committed:
bpo-29505: Fuzz json module, enforce size limit on int(x) fuzz (GH-13991)

* bpo-29505: Enable fuzz testing of the json module, enforce size limit on int(x) fuzz and json input size to avoid timeouts.

Contributed by by Ammar Askar for Google.

aixtools committed:
bpo-35545: Skip `test_asyncio.test_create_connection_ipv6_scope` on AIX (GH-14011)



because "getaddrinfo()" behaves different on AIX





https://bugs.python.org/issue35545

Commit:
bpo-37160: Thread native ID NetBSD support (GH-13835)

paulmon committed:
bpo-37201: fix test_distutils failures for Windows ARM64 (GH-13902)

paulmon committed:
bpo-37236: pragma optimize off for _Py_c_quot on Windows arm64 (GH-13983)

asvetlov committed:
Make asyncio stream sendfile fail on error (was hang) (GH-14025)

vstinner committed:
bpo-37223, test_io: silence last 'Exception ignored in:' (GH-14029)

Use catch_unraisable_exception() to ignore 'Exception ignored in:'
error when the internal BufferedWriter of the BufferedRWPair is
destroyed. The C implementation doesn't give access to the
internal BufferedWriter, so just ignore the warning instead.

vstinner committed:
bpo-37069: regrtest uses sys.unraisablehook (GH-13759)

regrtest now uses sys.unraisablehook() to mark a test as "environment
altered" (ENV_CHANGED) if it emits an "unraisable exception".
Moreover, regrtest logs a warning in this case.

Use "python3 -m test --fail-env-changed" to catch unraisable
exceptions in tests.

paulmon committed:
bpo-36779: time.tzname returns empty string on Windows if default cod… (GH-13073)



Calling setlocale(LC_CTYPE, "") on a system where GetACP() returns CP_UTF8 results in empty strings in _tzname[].

This causes time.tzname to be an empty string.
I have reported the bug to the UCRT team and will follow up, but it will take some time get a fix into production.

In the meantime one possible workaround is to temporarily change the locale by calling setlocale(LC_CTYPE, "C") before calling _tzset and restore the current locale after if the GetACP() == CP_UTF8 or CP_UTF7

@zooba 


https://bugs.python.org/issue36779

vstinner committed:
bpo-36402: Fix threading._shutdown() race condition (GH-13948)

Fix a race condition at Python shutdown when waiting for threads.
Wait until the Python thread state of all non-daemon threads get
deleted (join all non-daemon threads), rather than just wait until
Python threads complete.

* Add threading._shutdown_locks: set of Thread._tstate_lock locks
  of non-daemon threads used by _shutdown() to wait until all Python
  thread states get deleted. See Thread._set_tstate_lock().
* Add also threading._shutdown_locks_lock to protect access to
  threading._shutdown_locks.
* Add test_finalization_shutdown() test.

vstinner committed:
bpo-37253: Document PyCompilerFlags.cf_feature_version (GH-14019)

* Update PyCompilerFlags structure documentation.
* Document the new cf_feature_version field in the Changes in the C
  API section of the What's New in Python 3.8 doc.

vstinner committed:
bpo-37253: Add _PyCompilerFlags_INIT macro (GH-14018)

Add a new _PyCompilerFlags_INIT macro to initialize PyCompilerFlags
variables, rather than initializing cf_flags and cf_feature_version
explicitly in each variable.

vstinner committed:
bpo-37253: Fix typo in PyCompilerFlags doc (GH-14036)

Remove ";" to fix Sphinx formatting.

ned-deily committed:
Add 3.9 whatsnew file (GH-14040)

tim-one committed:
bpo-37257:  obmalloc:  stop simple arena thrashing (#14039)

GH-14039:  allow (no more than) one wholly empty arena on the usable_arenas list.

This prevents thrashing in some easily-provoked simple cases that could end up creating and destroying an arena on each loop iteration in client code.   Intuitively, if the only arena on the list becomes empty, it makes scant sense to give it back to the system unless we know we'll never need another free pool again before another arena frees a pool.  If the latter obtains, then - yes - this will "waste" an arena.

MakDon committed:
bpo-37216: update version to 3.9 in mac using document (GH-13966)

websurfer5 committed:
bpo-35070: test_getgrouplist may fail on macOS if too many groups (GH-13071)

vstinner committed:
bpo-37253: Remove PyAST_obj2mod_ex() function (GH-14020)

PyAST_obj2mod_ex() is similar to PyAST_obj2mod() with an additional
'feature_version' parameter which is unused.

jdemeyer committed:
bpo-37231: optimize calls of special methods (GH-13973)

vstinner committed:
tbpo-36402: Fix threading.Thread._stop() (GH-14047)

Remove the _tstate_lock from _shutdown_locks, don't remove None.

vstinner committed:
bpo-37210: Fix pure Python pickle when _pickle is unavailable (GH-14016)

Allow pure Python implementation of pickle to work
even when the C _pickle module is unavailable.

Fix test_pickle when _pickle is missing: declare PyPicklerHookTests
outside "if has_c_implementation:" block.

vstinner committed:
bpo-37261: Fix support.catch_unraisable_exception() (GH-14052)

The __exit__() method of test.support.catch_unraisable_exception
context manager now ignores unraisable exception raised when clearing
self.unraisable attribute.

MakDon committed:
bpo-6689: os.path.commonpath raises ValueError for different drives isn't documented (GH-14045)



It would raise ValueError("Paths don't have the same drive") if the paths on different drivers, which is not documented.


os.path.commonpath raises ValueError when the *paths* are in different drivers, but it is not documented.
Update the document according @Windsooon 's suggestion.
It actually raise ValueError according line 355 of [test of path](https://github.com/python/cpython/blob/master/Lib/test/test_ntpath.py) 


https://bugs.python.org/issue6689

pablogsal committed:
bpo-37213: Handle negative line deltas correctly in the peephole optimizer (GH-13969)

The peephole optimizer was not optimizing correctly bytecode after negative deltas were introduced. This is due to the fact that some special values (255) were being searched for in both instruction pointer delta and line number deltas.

vstinner committed:
bpo-36710: Pass explicitly tstate in sysmodule.c (GH-14060)

* Replace global var Py_VerboseFlag with interp->config.verbose.
* Add _PyErr_NoMemory(tstate) function.
* Add tstate parameter to _PyEval_SetCoroutineOriginTrackingDepth()
  and move the function to the internal API.
* Replace _PySys_InitMain(runtime, interp)
  with _PySys_InitMain(runtime, tstate).

aixtools committed:
bpo-37077: Add native thread ID (TID) for AIX (GH-13624)



This is the followup  for issue36084



https://bugs.python.org/issue37077

pablogsal committed:
bpo-37269: Correctly optimise conditionals with constant booleans (GH-14071)

Fix a regression introduced by af8646c8054d0f4180a2013383039b6a472f9698 that was causing code of the form:

if True and False:
   do_something()

to be optimized incorrectly, eliminating the block.

jdemeyer committed:
bpo-37249: add declaration of _PyObject_GetMethod (GH-14015)

vstinner committed:
bpo-37278: Fix test_asyncio ProactorLoopCtrlC (GH-14074)

Join the thread to prevent leaking a running thread and leaking a
reference.

Cleanup also the test:

* asyncioWindowsProactorEventLoopPolicy became the default policy,
  there is no need to set it manually.
* Only start the thread once the loop is running.
* Use a shorter sleep in the thread (100 ms rather than 1 sec).
* Use close_loop(loop) rather than loop.close().
* Use longer variable names.

geryogam committed:
Update concurrent.futures.rst (GH-14061)



This PR adds missing details in the [`concurrent.futures`](https://docs.python.org/3/library/concurrent.futures.html) documentation:

* the mention that `Future.cancel` also returns `False` if the call finished running;
* the mention of the states for `Future` that did not complete: pending or running.

asvetlov committed:
bpo-35998: Avoid TimeoutError in test_asyncio: test_start_tls_server_1() (GH-14080)

zooba committed:
Implement Windows release builds in Azure Pipelines (GH-14065)

ZackerySpytz committed:
bpo-19865: ctypes.create_unicode_buffer() supports non-BMP strings on Windows (GH-14081)

vstinner committed:
bpo-37261: Document sys.unraisablehook corner cases (GH-14059)

Document reference cycle and resurrected objects issues in
sys.unraisablehook() and threading.excepthook() documentation.

Fix test.support.catch_unraisable_exception(): __exit__() no longer
ignores unraisable exceptions.

Fix test_io test_writer_close_error_on_close(): use a second
catch_unraisable_exception() to catch the BufferedWriter unraisable
exception.

vstinner committed:
bpo-37266: Daemon threads are now denied in subinterpreters (GH-14049)

In a subinterpreter, spawning a daemon thread now raises an
exception. Daemon threads were never supported in subinterpreters.
Previously, the subinterpreter finalization crashed with a Pyton
fatal error if a daemon thread was still running.

* Add _thread._is_main_interpreter()
* threading.Thread.start() now raises RuntimeError if the thread is a
  daemon thread and the method is called from a subinterpreter.
* The _thread module now uses Argument Clinic for the new function.
* Use textwrap.dedent() in test_threading.SubinterpThreadingTests

vstinner committed:
bpo-35537: Rewrite setsid test for os.posix_spawn (GH-11721)

bpo-35537, bpo-35876: Fix also test_start_new_session() of
test_subprocess: use os.getsid() rather than os.getpgid().

vstinner committed:
Document C API changes in What's New in Python 3.8 (GH-14092)

zooba committed:
Fix Windows release build issues (GH-14091)

* Increase timeout for PGO builds in Windows release
* Fix test step failures
* Disable MinGW step properly
* Fix embeddable distro name

vstinner committed:
bpo-36707: Document "m" removal from sys.abiflags (GH-14090)

asvetlov committed:
bpo-37279: Fix asyncio sendfile support when  extra data are sent in fallback mode. (GH-14075)

asvetlov committed:
Use threadpool for reading from file in sendfile fallback mode (#14076)

geryogam committed:
Update weakref.rst (GH-14098)

ubordignon committed:
Fix typo in Lib/concurrent/futures/thread.py (GH-13953)

JulienPalard committed:
Doc: Bump Sphinx verison. (#13785)

To reflect the one we're using in production.

alex committed:
Update link in colorsys docs to be https (GH-14062)

JulienPalard committed:
Doc: Deprecation header: More precise wording. (GH-14109)

pablogsal committed:
bpo-37289: Remove 'if False' handling in the peephole optimizer (GH-14099)

aixtools committed:
bpo-28009: Fix uuid SkipUnless logic to be based on platform programs capable of introspection (GH-12777)

uuid could try fallback methods that had no chance of working on a particular
platform, and this could cause spurious test failures, as well as degraded
performance as fallback options were tried and failed.

This fixes both the uuid module and its test's SkipUnless logic to use a
prefiltered list of techniques that may at least potentially work on that platform.

Patch by Michael Felt (aixtools).

Commit:
bpo-35922: Fix RobotFileParser when robots.txt has no relevant crawl delay or request rate (GH-11791)

Co-Authored-By: Tal Einat <taleinat+github@gmail.com>

JulienPalard committed:
Doc: Remove an ugly space before a dot. (GH-14123)

mdickinson committed:
Simplify negativity checks in math.comb and math.perm. (GH-13870)

mdickinson committed:
Turn math.isqrt assertion into a comment to clarify its purpose. (GH-14131)

jdemeyer committed:
bpo-28805: document METH_FASTCALL (GH-14079)

shihai1991 committed:
closes bpo-37300: Remove unnecessary Py_XINCREF in classobject.c. (GH-14120)

terryjreedy committed:
bpo-37220: Fix 2.7 test -R crash on Windows. (GH-13957)

The patch needed for 2.7 should make the test more stable on 3.x also.

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

jsaporta committed:
Update 2019-06-17-01-37-25.bpo-29847.4VTssR.rst

jsaporta committed:
Added test for user-defined subclasses of Path; modified pathlib to make it pass.

jsaporta committed:
Merge branch 'fix-issue-29847' of github.com:jsaporta/cpython into fix-issue-29847

jsaporta committed:
Changed name of Path kwargs test.

jsaporta committed:
User subclass test now passing.
